core/src/forwarding_stage.rs

```diff
@@ -0,0 +1,396 @@
+    forward_address_getter: T,
+    connection_cache: Arc<ConnectionCache>,
+    data_budget: DataBudget,
+    udp_socket: UdpSocket,
+    metrics: ForwardingStageMetrics,
+}
+
+impl<T: ForwardAddressGetter> ForwardingStage<T> {
+    pub fn spawn(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        forward_address_getter: T,
+        connection_cache: Arc<ConnectionCache>,
+    ) -> JoinHandle<()> {
+        let forwarding_stage = Self::new(receiver, forward_address_getter, connection_cache);
+        Builder::new()
+            .name("solFwdStage".to_string())
+            .spawn(move || forwarding_stage.run())
+            .unwrap()
+    }
+
+    fn new(
+        receiver: Receiver<(BankingPacketBatch, bool)>,
+        forward_address_getter: T,
+        connection_cache: Arc<ConnectionCache>,
+    ) -> Self {
+        Self {
+            receiver,
+            forward_address_getter,
+            connection_cache,
+            data_budget: DataBudget::default(),
+            udp_socket: UdpSocket::bind("0.0.0.0:0").unwrap(),
+            metrics: ForwardingStageMetrics::default(),
+        }
+    }
+
+    fn run(mut self) {
+        while let Ok((packet_batches, tpu_vote_batch)) = self.receiver.recv() {
+            self.metrics.maybe_report();
+            self.handle_batches(packet_batches, tpu_vote_batch);
+        }
+    }
+
+    fn handle_batches(&mut self, packet_batches: BankingPacketBatch, tpu_vote_batch: bool) {
+        // Count the number of packets received
+        let num_packets_received: usize = packet_batches.0.iter().map(|batch| batch.len()).sum();
+        if tpu_vote_batch {
+            self.metrics.vote_packets_received += num_packets_received;
+        } else {
+            self.metrics.non_vote_packets_received += num_packets_received;
+        }
+
+        let Some(addr) = self
+            .forward_address_getter
+            .get_forwarding_address(tpu_vote_batch, self.connection_cache.protocol())
+        else {
+            return;
+        };
+
+        self.update_data_budget();
+        self.forward_batches(packet_batches, tpu_vote_batch, addr);
+    }
+
+    fn forward_batches(
+        &mut self,
+        packet_batches: BankingPacketBatch,
+        tpu_vote_batch: bool,
+        addr: SocketAddr,
+    ) {
+        let filtered_packets = packet_batches
+            .0
+            .iter()
+            .flat_map(|batch| batch.iter())
+            .filter(|p| !p.meta().discard())
+            .filter(|p| !p.meta().forwarded())
+            .filter(|p| p.meta().is_from_staked_node())
+            .filter(|p| self.data_budget.take(p.meta().size))
+            .filter_map(|p| p.data(..).map(|data| data.to_vec()));
+
+        let (success, num_packets) = if tpu_vote_batch {
+            let packets: Vec<_> = filtered_packets.into_iter().zip(repeat(addr)).collect();
+            let num_packets = packets.len();
+            let res = batch_send(&self.udp_socket, &packets);
```

Comment:
@lijunwangs just a heads up that this is coming and we'll need to add quic support for forwarded votes. This is still in draft though